crates/uv/src/commands/project/lock.rs

```diff
@@ -543,57 +543,84 @@ pub(crate) async fn read(workspace: &Workspace) -> Result<Option<Lock>, ProjectE
-    existing_lock: &Lock,
-    new_lock: &Lock,
-    workspace_root: &Path,
-    printer: Printer,
-) -> anyhow::Result<()> {
-    let existing_distributions: FxHashMap<PackageName, BTreeSet<Version>> =
+fn report_upgrades(existing_lock: &Lock, new_lock: &Lock, printer: Printer) -> anyhow::Result<()> {
+    let existing_distributions: FxHashMap<&PackageName, BTreeSet<&Version>> =
         existing_lock.distributions().iter().fold(
             FxHashMap::with_capacity_and_hasher(existing_lock.distributions().len(), FxBuildHasher),
             |mut acc, distribution| {
-                if let Ok(VersionId::NameVersion(name, version)) =
-                    distribution.version_id(workspace_root)
-                {
-                    acc.entry(name).or_default().insert(version);
-                }
+                acc.entry(distribution.name())
+                    .or_default()
+                    .insert(distribution.version());
                 acc
             },
         );
 
-    let new_distribution_names: FxHashMap<PackageName, BTreeSet<Version>> =
+    let new_distributions: FxHashMap<&PackageName, BTreeSet<&Version>> =
         new_lock.distributions().iter().fold(
             FxHashMap::with_capacity_and_hasher(new_lock.distributions().len(), FxBuildHasher),
             |mut acc, distribution| {
-                if let Ok(VersionId::NameVersion(name, version)) =
-                    distribution.version_id(workspace_root)
-                {
-                    acc.entry(name).or_default().insert(version);
-                }
+                acc.entry(distribution.name())
+                    .or_default()
+                    .insert(distribution.version());
                 acc
             },
         );
 
-    for (name, new_versions) in new_distribution_names {
-        if let Some(existing_versions) = existing_distributions.get(&name) {
-            if new_versions != *existing_versions {
+    for name in existing_distributions
+        .keys()
+        .chain(new_distributions.keys())
+        .collect::<BTreeSet<_>>()
+    {
+        match (
+            existing_distributions.get(name),
+            new_distributions.get(name),
+        ) {
+            (Some(existing_versions), Some(new_versions)) => {
+                if existing_versions != new_versions {
+                    let existing_versions = existing_versions
+                        .iter()
+                        .map(|version| format!("v{version}"))
+                        .collect::<Vec<_>>()
+                        .join(", ");
+                    let new_versions = new_versions
+                        .iter()
+                        .map(|version| format!("v{version}"))
+                        .collect::<Vec<_>>()
+                        .join(", ");
+                    writeln!(
+                        printer.stderr(),
+                        "{} {name} {existing_versions} -> {new_versions}",
+                        "Updating".green().bold()
```

Comment:
I think I may prefer "Updated", "Added", "Removed", etc. — I think we generally prefer "Did thing" over "Doing thing"?